### PR TITLE
Scalding people by splashing them with hot reagents

### LINKED ISSUE #33668: Hot reagents don't scald people
#### Description of issue
There's seemingly no mechanism for extremely hot liquids to scald people when splashed. One option is to just do an on-contact reaction thing that burns people based on temperature upon mob-reaction, similar to polyacid. Another is to have a "reagent coating/spatter" system like Dwarf Fortress has, and every Life() tick there could be thermal (and also chemical) interactions with the coated mob. Then you could also get wet people dripping onto the floor and stuff, which could be nice with puddles if they get implemented. A mob could have a single "covering" reagent datum, similar to how the blood/stomach reagents currently work, or they could have one per bodypart in the case of carbons. I'm curious to hear people's thoughts on this.

#### Steps to reproduce
Edit the temperature of the reagents in a container to thousands of degrees, then splash yourself with it.

### PR DESCRIPTION
**Pull request by Hinaichigo**
## What this does
This makes it so you can scald people by splashing them with hot reagents. Burn damage depends on both reagent temperature and volume. This table shows the damage scaling:

splash_damage_scaling

Burn damage is reduced by clothing covering the affected body part as long as the clothing has some degree of thermal protection, which also stacks eg. if you have both a jumpsuit and a suit-slot item covering the chest, their thermal protection will stack multiplicatively. Common clothing items often have 0.2-0.4 thermal conductivity which means only 20-40% of the damage gets through. Damage also can happen with very cold reagents, and is also burn damage.

Also, this makes it so making pills out of reagents equalizes the pill temperature with the temperature of the surrounding air, due to contact with the pilling machinery. This should help avoid burning people with pills of chems that require heated reactions to create.

Some considerations:
- the chem fridge gradually cools down its contents
- reagents can be cooled more rapidly by adding water via chem dispenser and then skimming off the water via chemmaster
- the "safe temperature range" for humans is 210 K to 360 K, or -63 C to ~87 C, so splashes within this range won't do any damage

Also I tweaked a few things with frying pan splashing/spilling text, generally making it a bit more concise, and also fixed a bug where the deep fryer sprite wasn't updating properly when sufficient reagent volume was removed from it via reagent container.

Closes  #33668

## Why it's good

- I think it adds more options and variability for combat. And it's a one-shot weapon you kind of have to prepare in-advance via the various reagent heating methods, and it also has a lot of variability in its damage output.
- It's a bit more realistic and you can also burn yourself a bit by drinking extremely hot liquids.
- It gives the chef a fun way to defend the kitchen and lead to scalding and slipping hijinks.
- There is potential for synergistic effects if spilling reagents when slipping while holding an open reagent container is re-implemented, especially if you can spill it on yourself, something I've been considering.
- For reagents to be eventually able to move around outside of reagent containers and mobs (puddles, fluids in pipes, coatings on mobs and objects), thermal interactions seem important to consider.

## Changelog
:cl:
 * rscadd: Getting splashed with hot liquids can lead to injury, be careful!
 * rscadd: Making pills with a ChemMaster equalizes the temperatures of the pills and the surrounding air.
 * tweak: Made frying pan splash text slightly more concise.
 * bugfix: Fixed deep fryer sprite not always updating properly when removing reagents.

### CODE REVIEW

**Comment by Eneocho** [comment]
"Now, listen carefully, coder... When you see the.... collabs... relay THESE words: Prepare for Unforeseen Consequences..."

**Comment by jwhitak** [comment]
How much damage does the laser scalpel-set temperature do on splash, with the highest thermal capacity chemical easily made from the dispenser per u used? Notable: 49u amounts (from a maximum filled smoke grenade)?

Reagent heating is a very bad system in the game right now, tolerated only because of the scalpel bug.

**Comment by DeityLink** [comment]
Hmm, can you use a lighter or scalpel to repeatedly heat up a beaker full of liquid iron until it's 5.000°C ? Or is there a cap to the various ways of raising reagent temperature?

**Comment by Hinaichigo** [comment]
There are caps on most ways of heating like bunsen burners or frying pans on a burning barrel or grill, which are a few hundred degrees in most cases.
The laser scalpel does also have a cap but it's in the thousands of degrees.
I suspect laser-scalpel temperature cap regents will do a massive amount of damage. I can check in a bit. We could set some kind of asymptotic effect on the damage if it's too powerful maybe. Or change the calculation in another way. Or now that bunsen burners work better we could maybe nerf laser scalpel heating, or implement some kind of ambient cooling where really hot things will slowly equalize temperature with the surrounding air. 

> Reagent heating is a very bad system in the game right now

I think this was fixed to an extent in #34281. Bunsen burners and putting a frying pan on the grill should heat things up more consistently now.

**Comment by jwhitak** [comment]
Heating notes:

Laser scalpels heat to 4500 instantly because they have heat_transfer set to 0. This makes the heat transfer function simply set the target to the source_temperature, which for laser scapels is TEMPERATURE_PLASMA = 4500. It ignores thermal mass in this case. A laser cautery on the other hand has a heat_transfer of 1500, so it adjusts it by adding (1500 / get_thermal_mass()) each click, capping at the TEMPERATURE_PLASMA. For very high thermal masses, this takes a *while*, but can be done comparably fast with macros spamming use.

The scalpel can be 'fixed' by setting heat_transfer to a > 0 number, but good luck passing that PR. Medbay gang will hire assassins.

**Comment by wotisjanitor** [comment]
The bunsen burner hasn't worked in like 2 years so if anyone nerfs the laser scalpel they will be hanged

**Comment by Hinaichigo** [comment]
The bunsen burner should be working now last time I tested it.

**Comment by DeityLink** [comment]
Imagine spess but with entropy and temperature calculations between atoms and the air surrounding them.

I'd code that if I didn't anticipate the ungodly lag that would cause.

**Comment by Hinaichigo** [comment]
Since most atoms don't have a "temperature" aside from mobs' body temperature, or anything with reagents in it, maybe it could work okay?
Maybe it could just be limited to `obj/item/weapon/reagent_containers` with nonzero reagent volume.
You could also store the thermal mass of the reagents on the reagents datum and only recalculate it on reagent change, and compare that with the heat capacity of the location's air.

**Comment by DeityLink** [comment]
> Since most atoms don't have a "temperature" aside from mobs' body temperature, or anything with reagents in it, maybe it could work okay?

To be credible, every atom *should* have a temperature, essentially the same way Oxygen Not Included handles it. Items could melt into slag puddles at high enough temperatures (in fact we DID have this feature at some point #1832. but it wasn't very engaging and caused plasma fires to be even more destructive)

**Comment by Hinaichigo** [comment]
I'm all for plasma fires being more destructive. At least they should melt inflatable shelters.

**Comment by jwhitak** [comment]
A beaker of water heated with laser scalpel does 380 fire damage. This PR isn't tested enough.

Edit: lasered fire extinguishers instantly crit people too 
Edit 2: Any chem that needs heat to make becomes lethally hot.
Edit 3: Hell cryo mixes instantly husk

**Comment by Dacendeth** [comment]
This is just insanely broken, You can spend 30 seconds heating up 5u of sugar in a beaker with a welding tool to make an instant kill 395 splash attack.

**Comment by Hinaichigo** [comment]
I think it could be reasonable to have an asymptotic cap around 30 damage or something like that? Or have sub-linear scaling?
Or is there a better way?
Please let me know any ideas on how to actually calculate the damage.

Current formula:
```
/mob/living/proc/get_burn_damage_from_thermal_transfer(thermal_energy_transferred, what_temp)
	return round(abs((TEMPERATURE_DAMAGE_COEFFICIENT / 1111111) * thermal_energy_transferred * get_safe_temperature_excursion(what_temp)))
```

Hopefully we can get something that feels fairly realistic but doesn't allow for one-click husking with just a little preparation.

**Comment by jwhitak** [comment]
Actually, I'm going to thumbs this down. There's no reliable method to cool chemicals right now, so this PR makes any chemical that requires heating lethal to use. Caramel, bone hurting juice, advanced burn chems, etc.

**Comment by Hinaichigo** [comment]
I think it could be reasonable to have an asymptotic cap around 30 damage or something like that? Or have sub-linear scaling?
Or is there a better way?
Please let me know any ideas on how to actually calculate the damage.

Current formula:
```
/mob/living/proc/get_burn_damage_from_thermal_transfer(thermal_energy_transferred, what_temp)
	return round(abs((TEMPERATURE_DAMAGE_COEFFICIENT / 1111111) * thermal_energy_transferred * get_safe_temperature_excursion(what_temp)))
```

Hopefully we can get something that feels fairly realistic but doesn't allow for one-click husking with only a little preparation.

**Comment by Hinaichigo** [comment]
There is one somewhat simple way to cool chems down actually. You can add room-temperature water to the beaker and then separate the water off again with the chemmaster after the temperatures equalizes.

**Comment by githubuser4141** [comment]
> This is just insanely broken, You can spend 30 seconds heating up 5u of sugar in a beaker with a welding tool to make an instant kill 395 splash attack.

could it be changed so that lights, scalpels and wtools can't heat up huge amounts of reagents in a beaker? like go out and hit your saucepan filled with cold water with a lighter, it's going to take a very long time and it probably won't reach boiling temp? Ghetto chemistry would remain because it'd depend on the size of the reagent pool AND the beaker size - so heating up 10 u of fluid in a coke can with a lighter to boiling temperature would be possible, heating up 100 u of reagents in a bluespace beaker - nope. (Should BS beakers even be heat-able? [at all] / [with ghetto tools?]

File: code/modules/mob/living/helper_procs.dm [hunk 1]
```diff
@@ -44,3 +44,40 @@ default behaviour is:
 /mob/living/proc/isDeadorDying()	//returns 1 if dead or in crit
 	if(stat == DEAD || health <= config.health_threshold_crit)
 		return TRUE
+
+/mob/living/proc/mob_thermal_mass()
+	//This can be fleshed out if it's used in more places. For now it's only used in calculating burn damage from reagents splashing a mob, so having this be a constant works alright.
+	return 2.98 * 70000 //specfic heat of 2.98 J / (g * degrees K), assume body mass of 70kg
```
Comment (west3436) [inline code/modules/mob/living/helper_procs.dm]: As you implied this is a nit, but the value could scale between fat, normal, and dwarf humans if you wanted to go into further detail here.

File: code/modules/reagents/Chemistry-Holder.dm [hunk 2]
```diff
@@ -582,30 +582,84 @@ trans_to_atmos(var/datum/gas_mixture/target, var/amount=1, var/multiplier=1, var
...
 
+#define SCALD_PAINFUL 15
+#define SCALD_AGONIZING 30
+
+/datum/reagents/proc/handle_reagents_mob_thermal_interaction(mob/living/L, method, list/zone_sels = ALL_LIMBS, volume_used)
+	if (!isliving(L))
+		return
+
+	var/ignore_thermal_prot = FALSE
+	if (method == INGEST) //Eating or drinking burns the mouth (head) regardless of targeting and isn't blocked by head thermal protection.
```
Comment (west3436) [inline code/modules/reagents/Chemistry-Holder.dm:618]: Perhaps stomach damage too?

Comment (Hinaichigo) [reply code/modules/reagents/Chemistry-Holder.dm:618]: As far as I know there's no stomach organ.
Dionae being able to eat weeds is based on their liver I think.

**Comment by Hinaichigo** [comment]
Apparently the high-precision laser scalpel doesn't heat to 4500 K instantly but the basic laser scalpel does.
I'm probably not going to nerf or mess with reagent heating in this PR, instead add some sublinear scaling here.
One thing is that burns from splashing (ingesting aside) to an extent are a "surface injury" so sublinear scaling makes sense.
I'm thinking that heat is transferred to the mob but generally this heat is absorbed by the large thermal bulk of the mob so their body temperature doesn't raise by that much (as seen in #33698 and #33891), instead it's mainly about the extreme temperatures that develop in the tissue during the brief period immediately following contact with the hot thing, the heat of which quickly dissipates to the surrounding tissue or atmosphere.

**Comment by Hinaichigo** [comment]
I changed the scaling to where I think it's decent. These are the raw numbers but keep in mind that they can be reduced by clothing. A grey jumpsuit for example has 60% damage reduction versus the numbers here:

Sipping the deep fryer oil (443 K):
~18 burn damage

Pouring a panful (100 u) of deep fryer oil onto your head:
~31 burn damage

Splashing yourself with a panful of laser scalpel temperature (4500 K) deep fryer oil:
~60 burn damage

Splashing yourself with a panful of 45000 K deep fryer oil:
~94 burn damage

Splashing someone with a large bluespace beaker (300 u) of 4500 K phalanximine
~93 burn damage

I can reduce it further if people want, but I also think we needn't be toooo conservative unless it really becomes something people use all the time; people might use this a lot at first but then as they get used to it it might use it less. After all you can crit someone via stungloves and then toolboxing their head all with materials you find at roundstart in tool storage. Or a chloral hydrate syringe stab into toolbox.

**Comment by Hinaichigo** [comment]
Even 300 u of 4500 K phalanximine on bare skin won't one-shot send someone to crit with this, but I do think maybe it would be good to wait on the other related PRs to see if they get enough approval. If not, we can consider if we want to nerf the burn damage a little bit here.

**Comment by Hinaichigo** [comment]
Gonna think about the scaling a bit on this.

**Comment by Hinaichigo** [comment]
Okay I updated the scaling to make it reasonable at household temperatures and not too broken at extreme temperatures. Check the OP to see a table.

**Comment by Hinaichigo** [comment]
@jwhitak I made the damage scaling nicer I think, and also used an idea originally suggested by Eneocho from #33891 where making pills equalizes the temperature of the pills with the surrounding air. Hopefully this will make burning people with medicine less of an issue.

**Comment by jwhitak** [comment]
> @jwhitak I made the damage scaling nicer I think, and also used an idea from #33891 where making pills equalizes the temperature of the pills with the surrounding air. Hopefully this will make burning people with medicine less of an issue.

you made a spreadsheet, based

Based on accurate data and use case scenarios. Now we know what it takes to husk people and it's more than one very large beaker of very hot stuff! Smoke will still destroy people from what I can see but that's more smoke's problem vs your code's problem.

The damage numbers are now reasonable, tested to the nearly unobtainable extremes. I am satisified and I will tell people to re-review this now that it has been tested(tm) with numbers in a speadsheet and the numbers don't instantly kill people with just a little bit of hot cola.

**Comment by Kurfursten** [comment]
I think this looks reasonable for merge. Conflicts, though.

**Comment by Hinaichigo** [comment]
Fixed conflicts.

**Comment by pedr69** [comment]
>rscadd: Making pills with a ChemMaster equalizes the temperatures of the pills and the surrounding air.

making pills out of anything laser-scalpeled will now instantly heat chemistry to 200C and drop all the firelocks

**Comment by Eneocho** [comment]
thank you hinaichigo very cool

**Comment by Hinaichigo** [comment]
Many of the anomalies like that and #34975 will be mitigated by #34855 if it is merged.
I can also add the heat-dissipation-upon-pilling to the "heat settings" admin panel toggle thing in that PR so that it'll only affect the air temperature if reagents are set to affect air temperature generally.

**Comment by pedr69** [comment]
i was about to make a bugreport that the reagent cooling was broken
turns out it hasn't been merged yet

**Pull request merged**